src/java.base/share/classes/java/nio/Bits.java

```diff
@@ -107,71 +107,61 @@ static boolean unaligned() {
+        // there is no need to rush.
+        //
+        // If this code is modified, make sure a stress test like DirectBufferAllocTest
+        // performs well.
+
         boolean interrupted = false;
         try {
+            BufferCleaner.Canary canary = null;
 
-            // Retry allocation until success or there are no more
-            // references (including Cleaners that might free direct
-            // buffer memory) to process and allocation still fails.
-            boolean refprocActive;
-            do {
-                try {
-                    refprocActive = jlra.waitForReferenceProcessing();
-                } catch (InterruptedException e) {
-                    // Defer interrupts and keep trying.
-                    interrupted = true;
-                    refprocActive = true;
-                }
-                if (tryReserveMemory(size, cap)) {
-                    return;
-                }
-            } while (refprocActive);
-
-            // trigger VM's Reference processing
-            System.gc();
-
-            // A retry loop with exponential back-off delays.
-            // Sometimes it would suffice to give up once reference
-            // processing is complete.  But if there are many threads
-            // competing for memory, this gives more opportunities for
-            // any given thread to make progress.  In particular, this
-            // seems to be enough for a stress test like
-            // DirectBufferAllocTest to (usually) succeed, while
-            // without it that test likely fails.  Since failure here
-            // ends in OOME, there's no need to hurry.
             long sleepTime = 1;
-            int sleeps = 0;
-            while (true) {
+            for (int sleeps = 0; sleeps < MAX_SLEEPS; sleeps++) {
+                // See if we can satisfy the allocation now.
                 if (tryReserveMemory(size, cap)) {
                     return;
                 }
-                if (sleeps >= MAX_SLEEPS) {
-                    break;
+
+                if (canary == null || canary.isDead()) {
+                    // If canary is not yet initialized, we have not triggered a cleanup.
+                    // If canary is dead, there was progress, and it was not enough.
+                    // Trigger GC to perform reference processing and then cleaning.
+                    canary = BufferCleaner.newCanary();
+                    System.gc();
                 }
+
+                // Exponentially back off waiting for Cleaner to catch up.
                 try {
-                    if (!jlra.waitForReferenceProcessing()) {
-                        Thread.sleep(sleepTime);
-                        sleepTime <<= 1;
-                        sleeps++;
-                    }
+                    Thread.sleep(sleepTime);
+                    sleepTime *= 2;
                 } catch (InterruptedException e) {
                     interrupted = true;
+                    break;
```

Comment:
I'm not sure that bailing out to throw OOME when interrupted is the right thing to do here. If ByteBuffer.allocateDIrect is called with the interrupt status set, or someone interrupts a thread while allocating on this slow path, then it would be surprising to have it throw OOME when it might have succeeded (with the interrupt status set).

Reply:
Agreed. I thought I maintained the same interrupt behavior like before, but now I see old code does not bail on interrupt, it just restores the interrupt status on both failing and passing paths. Dropped this `break` in new commit.